src/security/Session.cc

```diff
@@ -33,12 +35,27 @@ static int
 tls_read_method(int fd, char *buf, int len)
 {
     auto session = fd_table[fd].ssl.get();
-    debugs(83, 3, "started for session=" << (void*)session);
+    debugs(83, 5, "started for session=" << static_cast<void*>(session) << " FD " << fd << " buf.len=" << len);
+
+    Security::PrepForIo();
 
 #if USE_OPENSSL
     int i = SSL_read(session, buf, len);
+    const auto savedErrno = errno; // zero if SSL_read() does not set it
+
+    if (i <= 0) {
+        debugs(83, 3, "SSL_read(FD " << fd << ") error(" << i << "): " << SSL_get_error(session, i) << ReportSysError(savedErrno));
+        Security::ForgetErrors(); // will debugs() errors before forgetting them
+        errno = savedErrno;
+    }
 #elif HAVE_LIBGNUTLS
     int i = gnutls_record_recv(session, buf, len);
+    const auto savedErrno = errno; // zero if gnutls_record_recv() does not set it
+
+    if (i < 0) {
```

Comment:
We use `i <= 0` in OpenSSL code but use `i < 0` in GnuTLS code. Both are correct AFAICT because the two libraries treat zero return value differently: Only OpenSSL uses zero returns to flag certain errors.

This comment does not request and PR changes.